kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/ModemTaskScheduler.java

```diff
@@ -0,0 +1,191 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Eurotech and/or its affiliates and others
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Eurotech
+ *******************************************************************************/
+package org.eclipse.kura.nm;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.freedesktop.dbus.exceptions.DBusException;
+import org.freedesktop.dbus.exceptions.DBusExecutionException;
+import org.freedesktop.modemmanager1.Modem;
+import org.freedesktop.networkmanager.Device;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ModemTaskScheduler {
+
+    private static final Logger logger = LoggerFactory.getLogger(ModemTaskScheduler.class);
+
+    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(3);
+    private final Device device;
+    private final int maxFail;
+    private final int holdoff;
+    private final boolean autoconnect;
+    private final int resetDelayMinutes;
+    private final String deviceId;
+
+    private Optional<NMDbusConnector> nmDbusConnector = Optional.empty();
+    private ScheduledFuture<?> connectionHandler;
+    private ScheduledFuture<?> resetHandler;
+    private AtomicBoolean isConnectionScheduled = new AtomicBoolean(false);
+    private AtomicBoolean isResetScheduled = new AtomicBoolean(false);
+    private int delay = 0;
+
+    public ModemTaskScheduler(String deviceId, Device device, NetworkProperties properties) {
+
+        try {
+            this.nmDbusConnector = Optional.of(NMDbusConnector.getInstance());
+        } catch (DBusExecutionException | DBusException e) {
+            logger.error("Cannot initialize NMDbusConnector due to: ", e);
+        }
```

Comment:
Is this really optional though? From the functionality standpoint, the scheduler cannot work properly without a reference to the `NMDbusConnector`. Is this due to lifecycle issues? Can the `ModemTaskScheduler` be instantiated when `NMDbusConnector` is not?

Reply:
From the functionality point of view, indeed the `ModemTaskScheduler` will not work. However, if the `NMDbusConnector` cannot be retrieved and it throws a `DBusException`, it will be null generating a lot of NPE. So, I preferred to make it optional and log if it is not present.